compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/DecomposeIm2col.cpp

```diff
@@ -12,36 +12,56 @@
 #include "mlir/Dialect/MemRef/Transforms/Transforms.h"
 #include "mlir/Dialect/SCF/IR/SCF.h"
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 
 namespace mlir::iree_compiler::IREE::LinalgExt {
 
 #define GEN_PASS_DEF_DECOMPOSEIM2COLPASS
 #include "iree/compiler/Dialect/LinalgExt/Transforms/Passes.h.inc"
 
-namespace {
-
-/// Pattern to decompose the tiled im2col op.
-struct DecomposeIm2col : public OpRewritePattern<Im2colOp> {
-  using OpRewritePattern<Im2colOp>::OpRewritePattern;
+static LogicalResult decomposeIm2col(Im2colOp im2colOp, RewriterBase &rewriter,
+                                     bool unroll) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(im2colOp);
+  FailureOr<SmallVector<Value>> decomposedIm2col =
+      im2colOp.decomposeOperation(rewriter);
+  if (failed(decomposedIm2col)) {
+    return failure();
+  }
+  rewriter.replaceOp(im2colOp, decomposedIm2col.value().front());
+  if (!unroll) {
+    return success();
+  }
 
-  LogicalResult matchAndRewrite(Im2colOp im2colOp,
-                                PatternRewriter &rewriter) const override {
-    FailureOr<SmallVector<Value>> decomposedIm2col =
-        im2colOp.decomposeOperation(rewriter);
-    if (failed(decomposedIm2col)) {
+  // Unroll the loop nest created by the im2col op decomposition.
+  auto outerLoop = decomposedIm2col.value().front().getDefiningOp<scf::ForOp>();
```

Comment:
nit: Add an assert that `outerLoop` exists as a marker in case someone tries to change the decomposition.